### ISSUE TITLE
Declare subpath exports for bundled plugins, Eleventy-maintained data, etc.

### ISSUE THREAD

**Issue by noelforte** [11ty/eleventy#3460]
### Is your feature request related to a problem? Please describe.

Upfront note: I'd be willing to submit a PR to help out with this.

Commit c505822 inadvertantly introduced some breaking changes to [subpath exports](https://nodejs.org/api/packages.html#subpath-exports) when Eleventy is consumed as a dependency. By adding the `exports` key to package.json with only value, any attempt to import a subpath throws an error.

### Describe the solution you'd like

This is more specific to `@11ty/eleventy/src/Engines/Util/ContextAugmenter.js` (https://github.com/11ty/eleventy/issues/3355#issuecomment-2316378398) but is applicable to any `exports` in Eleventy that **could** be consumed by outside packages, including (but not limited to):

- Plugins bundled with Eleventy
- `.d.ts` declaration files ([#3097](https://github.com/11ty/eleventy/issues/3097), [#3296](https://github.com/11ty/eleventy/issues/3296))
- Eleventy-maintained global data and methods like `augmentKeys`

Syntax I'm thinking of could be something like:

```json
"exports": {
  ".": {
    "import": "./src/Eleventy.js",
    "require": "./src/EleventyCommonJs.cjs",
  },
  "./user-config": {
    "types": "./types/UserConfig.d.ts"
  },
  "./plugins/html-base": "./src/Plugins/HtmlBasePlugin.js",
  "./plugins/render": "./src/RenderPlugin.js",
  "./plugins/input-path-to-url": "./src/InputPathToUrl.js",
  "./utils/context-augmentor": "./src/Engines/Util/ContextAugmenter.js"
}
```

The above declarations can be consumed as so:

```ts
import type UserConfig from '@11ty/eleventy/user-config';
import Eleventy from '@11ty/eleventy';
import HtmlBasePlugin from '@11ty/eleventy/plugins/html-base';
// ...
```

Or via an aggregator:

```json
"exports": {
  "./internal-plugins": "./src/Plugins/index.js",
}
```

```js
import {
  HtmlBasePlugin,
  RenderPlugin,
  InputPathToUrlPlugin,
} from '@11ty/eleventy/internal-plugins';
```

### Additional context

Subpath exports are supported by Node 18 which is the required minimum for 3.0. Additionally, when an `exports` field is present the `main` field isn't looked at by Node.js except in the case of `require()` statements, but with a `exports[path].require` field it's redundant to have for Node 18+.

**Comment by Ryuno-Ki** [comment]
Related to #3097

**Comment by noelforte** [comment]
> Related to #3097

@Ryuno-Ki yep, I mentioned that issue in my original post. I opened a new issue since while .d.ts files can be referenced through subpath imports, the implementation of subpath imports isn't _directly_ dependent on typings. I'm ok closing in favor of the other issue, but there's my explanation 😄

**Comment by Ryuno-Ki** [comment]
Ah, missed the nuance.

**Comment by stevenvachon** [comment]
Importing the types still doesn't work because `"types"` is supposed to be automatically loaded when loading either `"import"` or `"require"`, but neither of those are defined currently:

```json
	"exports": {
		".": {
			"import": "./src/Eleventy.js",
			"require": "./src/EleventyCommonJs.cjs"
		},
		"./UserConfig": {
			"types": "./src/UserConfig.js"
		}
	},
```

Doing either of these imports nothing:
```ts
import type UserConfig from '@11ty/eleventy/UserConfig';
```
```ts
import UserConfig from '@11ty/eleventy/UserConfig';
```

**Comment by noelforte** [comment]
@stevenvachon This issue is about subpath exports, not typing support from Eleventy. Subpath and conditional exports help facilitate exposing types but only where the codebase supports it.

> Importing the types still doesn't work because `"types"` is supposed to be automatically loaded when loading either `"import"` or `"require"`, but neither of those are defined currently:
>   ```json
> 	"exports": {
> 		".": {
> 			"import": "./src/Eleventy.js",
> 			"require": "./src/EleventyCommonJs.cjs"
> 		},
> 		"./UserConfig": {
> 			"types": "./src/UserConfig.js"
> 		}
> 	},
>   ```

This code is correct, since "types" is actually a non-standard specifier in `package.json`, but does get analyzed by TypeScript and by extension VSCode and other IDE's when a module is imported. You can read the specification on [Subpath Exports](https://nodejs.org/docs/latest-v22.x/api/packages.html#subpath-exports) and then a little farther down that page the documentation for [Conditional Exports](https://nodejs.org/docs/latest-v22.x/api/packages.html#conditional-exports). 

The export specifiers are mutually exclusive, but if used together the path rule **must** be used as a superset of the condition. In this case, UserConfig is only being used for types, and the root export provides both an import and require specifier.

> Doing either of these imports nothing:
> 
> import type UserConfig from '@11ty/eleventy/UserConfig';
> 
> import UserConfig from '@11ty/eleventy/UserConfig';

Eleventy is written in pure JS and does not provide `.d.ts` files currently making types difficult to pick up from TypeScript. Nevertheless, a few options:

1. Don't write your Eleventy config in Typescript, and instead use JSDoc-style type declarations to get type hints:
    ```js
    /** @param {import('@11ty/eleventy/UserConfig')} eleventyConfig */
    export default async function eleventy(eleventyConfig) {
      // your config here
    }
    ```

2. Increase `compilerOptions.maxNodeModuleJsDepth` to `2` in `tsconfig.json`. Since UserConfig is buried a level deeper, typescript might have a hard time finding it. (I've had mixed results with this, most recently setting it to `3` did the trick... 😵‍💫)
3. Use the excellent 11ty.ts plugin by @panoply to get the config types and/or wrap your config function with a `defineConfig` provided by that plugin.
